mypy/checkexpr.py

```diff
@@ -2076,60 +2081,106 @@ def visit_reveal_expr(self, expr: RevealExpr) -> Type:
             return AnyType(TypeOfAny.from_another_any, source_any=tp)
         return AnyType(TypeOfAny.special_form)
 
     def visit_type_alias_expr(self, alias: TypeAliasExpr) -> Type:
-        """Get type of a type alias (could be generic) in a runtime expression."""
-        if isinstance(alias.type, Instance) and alias.type.invalid:
+        """Right hand side of a type alias definition.
+
+        It has the same type as if the alias itself was used in a runtime context.
+        For example:
+
+            A = reveal_type(List[T])
+            reveal_type(A)
+
+        will reveal the same type `def (...) -> builtins.list[Any]`.
+        """
+        return self.alias_type_in_runtime(alias.type, alias.tvars, alias.no_args, alias)
+
+    def alias_type_in_runtime(self, target: Type, alias_tvars: List[str],
+                              no_args: bool, ctx: Context) -> Type:
+        """Get type of a type alias (could be generic) in a runtime expression.
+
+        Note that this function can be called only if the alias appears _not_
+        as a target of type application, which is treated separately in the
+        visit_type_application method.
+        """
+        if isinstance(target, Instance) and target.invalid:
             # An invalid alias, error already has been reported
             return AnyType(TypeOfAny.from_error)
-        item = alias.type
-        if not alias.in_runtime:
-            # We don't replace TypeVar's with Any for alias used as Alias[T](42).
-            item = set_any_tvars(item, alias.tvars, alias.line, alias.column)
+        # If this is a generic alias, we set all variables to `Any`.
+        # For example:
+        #     A = List[Tuple[T, T]]
+        #     x = A() <- same as List[Tuple[Any, Any]], see PEP 484.
+        item = set_any_tvars(target, alias_tvars, ctx.line, ctx.column)
         if isinstance(item, Instance):
             # Normally we get a callable type (or overloaded) with .is_type_obj() true
             # representing the class's constructor
             tp = type_object_type(item.type, self.named_type)
+            if no_args:
+                return tp
+            return self.safe_apply(tp, item.args, ctx)
         else:
             # This type is invalid in most runtime contexts
             # and corresponding an error will be reported.
-            return alias.fallback
+            kind = (' to Callable' if isinstance(item, CallableType) else
+                    ' to Tuple' if isinstance(item, TupleType) else
+                    ' to Union' if isinstance(item, UnionType) else '')
+            cdef = ClassDef('Type alias' + kind, Block([]))
+            fb_info = TypeInfo(SymbolTable(), cdef, self.chk.tree.fullname())
+            fb_info.bases = [self.object_type()]
+            fb_info.mro = [fb_info, self.object_type().type]
+            return Instance(fb_info, [])
+
+    def safe_apply(self, tp: Type, args: List[Type], ctx: Context) -> Type:
```

Comment:
> Why is that not an error?

Because `type_object_type` could return other types (at least `Any`). But in that cases the error should be already reported. I will add a comment.